networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/gossip/subnets/NodeBasedStableSubnetSubscriber.java

```diff
@@ -0,0 +1,122 @@
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.networking.eth2.gossip.subnets;
+
+import static java.util.Collections.emptySet;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NavigableSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.units.bigints.UInt256;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.validator.SubnetSubscription;
+
+public class NodeBasedStableSubnetSubscriber implements StableSubnetSubscriber {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final AttestationTopicSubscriber persistentSubnetSubscriber;
+  private final NavigableSet<SubnetSubscription> subnetSubscriptions =
+      new TreeSet<>(
+          Comparator.comparing(SubnetSubscription::getUnsubscriptionSlot)
+              .thenComparing(SubnetSubscription::getSubnetId));
+  private final Spec spec;
+  private final int subnetsPerNode;
+  private final Optional<UInt256> discoveryNodeId;
+
+  public NodeBasedStableSubnetSubscriber(
+      final AttestationTopicSubscriber persistentSubnetSubscriber,
+      final Spec spec,
+      final Optional<UInt256> discoveryNodeId) {
+    this.persistentSubnetSubscriber = persistentSubnetSubscriber;
+    this.spec = spec;
+    this.subnetsPerNode = spec.getNetworkingConfig().getSubnetsPerNode();
+    this.discoveryNodeId = discoveryNodeId;
+  }
+
+  @Override
+  public void onSlot(final UInt64 slot) {
+    // Iterate through current subscriptions to remove the ones that have expired
+    final Iterator<SubnetSubscription> iterator = subnetSubscriptions.iterator();
+    while (iterator.hasNext()) {
+      final SubnetSubscription subnetSubscription = iterator.next();
+      if (subnetSubscription.getUnsubscriptionSlot().compareTo(slot) > 0) {
+        break;
+      }
+      iterator.remove();
+    }
+
+    // Adjust the number of subscriptions
+    // If there are new subscriptions, pass the new subscription set to BeaconNode
+    Set<SubnetSubscription> newSubnetSubscriptions =
+        adjustNumberOfSubscriptionsToNodeRequirement(slot);
+    if (!newSubnetSubscriptions.isEmpty()) {
+      persistentSubnetSubscriber.subscribeToPersistentSubnets(newSubnetSubscriptions);
+    }
+  }
+
+  /**
+   * Adjusts the number of subscriptions to SUBNETS_PER_NODE. Returns the set of new subscriptions
+   * that were added, if there were no new subscriptions, it returns an empty set.
+   */
+  private Set<SubnetSubscription> adjustNumberOfSubscriptionsToNodeRequirement(
+      final UInt64 currentSlot) {
+
+    if (subnetSubscriptions.size() >= subnetsPerNode) {
+      return emptySet();
+    }
+    LOG.info(
```

Comment:
I'm not sure we need this in info. Plus if everything is correct it will log not very useful info - "from 0 to 2" always. I'd add instead DEBUG logging on subscription/unsubscription, something like Subscribed to persistent subnet 12. Unsubscribed from persistent  subnet 45.

Reply:
Done. Added subscription and un-subscription INFO log. Will replace them by DEBUG after testing